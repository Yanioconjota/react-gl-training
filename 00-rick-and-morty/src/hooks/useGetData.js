import { useState, useEffect } from "react";
import { getData } from "../helpers/getData";

export const useGetData = () => {
  
  const [state, setState] = useState({
    characters: [],
    loading: true
  });

  useEffect(() => {
    getData()
      .then(characters => {
        setState({
          characters: characters,
          loading: false,
        });
      })
  }, []);
  
  return state;
};
