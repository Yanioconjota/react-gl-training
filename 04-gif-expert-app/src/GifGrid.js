import React, { useState, useEffect } from "react";
import { Col, Row } from "react-bootstrap";
import GifItem from "./GifItem";

const GifGrid = ({ category }) => {

  const [images, setImages] = useState([]);

  useEffect(() => {
    getGifs(category);
  }, [category]);

  const getGifs = async(category) => {
    const apiKey = "api_key=EJRMW0vkab9582ndI3I5UZQ4ObyWb3OK";
    const query = category;
    const url = `https://api.giphy.com/v1/gifs/search?q=${query}&limit=12&${apiKey}`
    const resp = await fetch(url);
    const {data} = await resp.json();
    
    const gifs = data.map((img) => ({
      id: img.id,
      title: img.title,
      url: img.images?.downsized_large.url,
    }));

    console.log(gifs);
    setImages(gifs);
  };

  return (
    <>
      <h3>{category}</h3>
      <Row>
        {images.map((img) => (
          <Col xs={12} md={3} key={img.id}>
            <GifItem {...img}/>
          </Col>
        ))}
      </Row>
    </>
  );
}

export default GifGrid;