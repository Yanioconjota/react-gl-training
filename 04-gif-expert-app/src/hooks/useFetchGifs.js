import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";


export const useFetchGifs = (category, limit, offset) => {

  const [state, setState] = useState({
    data: [],
    loading: true,
    limit: limit,
    offset: offset,
  });

  useEffect(() => {
    getGifs(category, limit, offset)
      .then( imgs => {
        setState({
          data: imgs,
          loading: false,
          limit: limit,
          offset: offset 
        });
      });
  }, [category, limit, offset]);

  return state; //{ data: [], loading: true }
}