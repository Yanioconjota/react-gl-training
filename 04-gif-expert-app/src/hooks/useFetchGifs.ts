import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";
import { IGifState } from "../modules/exports";


export const useFetchGifs = (category: string, limit: number, offset: number) => {

  const [state, setState] = useState<IGifState>({
    data: [],
    loading: true,
    limit: limit,
    offset: offset,
  });

  useEffect(() => {
    getGifs(category, limit, offset)
      .then( imgs => {
        setState({
          data: imgs,
          loading: false,
          limit: limit,
          offset: offset 
        });
      });
  }, [category, limit, offset]);

  return state; //{ data: [], loading: true }
}