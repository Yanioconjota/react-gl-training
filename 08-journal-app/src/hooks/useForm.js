import { useState } from "react";

export const useForm = (initialState = {}) => {
  //Se pueden agregar las validaciones del formulario acÃ¡
  const [values, setValues] = useState(initialState);

  const handleInputChange = ({ target }) => {
    setValues({
      ...values,
      [target.name]: target.value,
    });
  };

  const reset = () => {
    setValues(initialState);
  };

  return [values, handleInputChange, reset];
};
