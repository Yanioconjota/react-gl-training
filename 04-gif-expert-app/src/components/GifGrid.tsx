import React, { useState } from "react";
import { Col, Row, Button } from "react-bootstrap";
import GifItem from "./GifItem";
import { useFetchGifs } from "../hooks/useFetchGifs";
import PropTypes from 'prop-types';

const GifGrid = ({ category }) => {

  const multiplier = 12;

  const [state, setState] = useState({
    limit: multiplier,
    offset: 0
  });

  const { data: images, loading } = useFetchGifs(category, state.limit, state.offset);

  const loadMore = () => {
    setState({
      limit: state.limit + multiplier,
      offset: state.offset + 1
    })
  };

  return (
    <>
      { loading ? <p className="animate__animated animate__flash">Loading...</p> :
        <>
          <h3 className="animate__animated animate__fadeIn">{category}</h3>
          <Row>
            {images.map((img) => (
              <Col xs={12} md={3} key={img.id}>
                <GifItem {...img} />
              </Col>
            ))}
            <Button variant="primary" onClick={loadMore}>Load More</Button>
          </Row>
        </>
      }
    </>
  );
}

GifGrid.propTypes = {
  category: PropTypes.arrayOf(PropTypes.string).isRequired
}

export default GifGrid;