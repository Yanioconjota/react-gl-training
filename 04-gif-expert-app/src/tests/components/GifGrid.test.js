import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import GifGrid from "../../components/GifGrid";
import { useFetchGifs } from "../../hooks/useFetchGifs";
//Mock a la llamada al método useFetchGifs
jest.mock("../../hooks/useFetchGifs");
describe('pruebas con gifGrid', () => {
  const category = "Far Cry 6";
  test('El componente debe de mostrarse correctamente', () => {
    useFetchGifs.mockReturnValue({
      data: [],
      loading: true
    });
    
    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de mostrar items cuando se cargan las imágenes useFetch', () => {
    const imgs = [
      {
        id: "0d1TtPiMfn8q68dhtU",
        title: "Danny Trejo Cooking Tacos",
        url: "https://localhost/cualquiercosa",
      },
      {
        id: "0d1TtPiMfn8q68dhts",
        title: "Danny Trejo Cooking Tacos again",
        url: "https://localhost/cualquiercosa",
      },
    ];
    useFetchGifs.mockReturnValue({
      data: imgs,
      loading: false,
      limit: 12,
      offset: 0,
    });
    const wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("p").exists()).toBe(false);
    expect(wrapper.find("GifItem").length).toBe(imgs.length);
  });
});