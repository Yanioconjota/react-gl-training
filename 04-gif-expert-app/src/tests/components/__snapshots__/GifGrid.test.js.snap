// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas con gifGrid El componente debe de mostrarse correctamente 1`] = `
<Fragment>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
</Fragment>
`;

exports[`pruebas con gifGrid debe de mostrar items cuando se cargan las im√°genes useFetch 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Far Cry 6
  </h3>
  <Row>
    <Col
      key="0d1TtPiMfn8q68dhtU"
      md={3}
      xs={12}
    >
      <GifItem
        id="0d1TtPiMfn8q68dhtU"
        title="Danny Trejo Cooking Tacos"
        url="https://localhost/cualquiercosa"
      />
    </Col>
    <Col
      key="0d1TtPiMfn8q68dhts"
      md={3}
      xs={12}
    >
      <GifItem
        id="0d1TtPiMfn8q68dhts"
        title="Danny Trejo Cooking Tacos again"
        url="https://localhost/cualquiercosa"
      />
    </Col>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      variant="primary"
    >
      Load More
    </Button>
  </Row>
</Fragment>
`;
